backend/notification/rate_limiter.js

```diff
@@ -1,34 +1,119 @@
+const prisma = require("../prisma/prisma_client");
+
 class RateLimiter {
-  constructor(time_window, maximum_requests) {
-    this.time_window = time_window * 1000;
-    this.maximum_requests = maximum_requests;
+  constructor(timeWindow, maximumRequests) {
+    this.timeWindow = timeWindow * 1000;
+    this.maximumRequests = maximumRequests;
     this.userRequestMap = new Map();
   }
 
-  useRateLimiter(userId) {
+  async saveNotificationInteraction(userId, category, notificationId) {
+    await prisma.notificationInteractions.create({
+      data: {
+        userId: userId,
+        timestamp: new Date(),
+        category: category,
+        notificationId: notificationId,
+      },
+    });
+  }
+
+  async getNotificationInteractionsData(userId) {
+    const interactions = await prisma.notificationInteractions.findMany({
+      where: {
+        userId: userId,
+      },
+      orderBy: {
+        timestamp: "desc",
+      },
+    });
+    return interactions;
+  }
+
+  async getUserMostActiveTimeWIndows(userId) {
+    const interactions = await this.getNotificationInteractionsData(userId);
+    const hourCounts = new Array(24).fill(0);
+    interactions.forEach((item) => {
+      const hour = new Date(item.timestamp).getHours();
+      hourCounts[hour]++;
+    });
+
+    const activeHours = hourCounts.map((numberOfClicks, hourIndex) => ({
+      hourIndex,
+      numberOfClicks,
+    }));
+    return activeHours
+      .sort((a, b) => b.numberOfClicks - a.numberOfClicks)
+      .slice(0, 3); //Returns top user's top 3 active hours
+  }
+
+  async getUserPreferredNotificationCategory(userId) {
+    const interactions = await this.getNotificationInteractionsData(userId);
+    const categoriesCount = {};
+    interactions.forEach((item) => {
+      categoriesCount[item.category] =
+        (categoriesCount[item.category] || 0) + 1;
+    });
+    const sortedCategories = Object.entries(categoriesCount).sort(
+      (a, b) => b[1] - a[1]
+    );
+    return sortedCategories[0][0]; //Returns most preffered category
+  }
+
+  async useRateLimiter(userId, category) {
     const currentTime = Date.now();
+    const currentHour = new Date().getHours();
+    const userMostActiveHours = await this.getUserMostActiveTimeWIndows(userId);
+    const userActiveHours = userMostActiveHours.filter(
+      (item) => item.hourIndex === currentHour
+    );
+    const isUserActiveTimeWindow = userActiveHours.length > 0;
```

Comment:
If you follow the recommendation above you would just need to check if the current hour is in the array of most active hours

Reply:
Thanks, Fixed